### PR TITLE
suggestions from disk index

### PR DESCRIPTION
**Pull request by sokrahta**
addresses #68
suggestion time increased to 175% of previous, which is acceptable.
index time memory allocations dropped to 59% of previous, and with the index being stored on disk it should fade out of active memory naturally over time.

benchmark summaries:
all expenses paid
```
[info] Benchmark                          Mode  Cnt     Score     Error  Units
[info] ESSuggestBenchmark.suggest         avgt   16  2087.518 ± 161.728  ms/op
```
without doc count query
```
[info] Benchmark                                                                Mode  Cnt    Score    Error  Units
[info] c.s.s.common.CompletionAnalyzerBenchmark.analyzeCompletionTokens         avgt   16  115.608 ±  4.350  us/op
[info] c.s.s.secondary.ESIndexBenchmark.indexKeyword                            avgt   16  110.553 ± 20.704  ms/op
[info] c.s.s.secondary.ESIndexBenchmark.indexPhrase                             avgt   16  148.281 ±  3.929  ms/op

[info] ESIndexBenchmark.indexKeyword                                   avgt   16         99.034 ±       22.399   ms/op
[info] ESIndexBenchmark.indexKeyword:·gc.alloc.rate                    avgt   16        529.941 ±       91.756  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.alloc.rate.norm               avgt   16  130277025.664 ±  1604773.826    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Eden_Space           avgt   16        512.563 ±      318.465  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Eden_Space.norm      avgt   16  116641849.200 ± 66445648.071    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Old_Gen              avgt   16          0.145 ±        0.263  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Old_Gen.norm         avgt   16      40004.777 ±    79728.025    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Perm_Gen             avgt   16         ≈ 10⁻⁴                 MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Perm_Gen.norm        avgt   16         18.744 ±       15.916    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Survivor_Space       avgt   16          4.439 ±        1.258  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Survivor_Space.norm  avgt   16    1145163.249 ±   439025.664    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.count                         avgt   16        139.000                 counts
[info] ESIndexBenchmark.indexKeyword:·gc.time                          avgt   16       4076.000                     ms
[info] ESIndexBenchmark.indexPhrase                                    avgt   16        125.336 ±       11.185   ms/op
[info] ESIndexBenchmark.indexPhrase:·gc.alloc.rate                     avgt   16        759.892 ±       63.986  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.alloc.rate.norm                avgt   16  242982345.605 ±  2087913.884    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Eden_Space            avgt   16        799.454 ±       28.358  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Eden_Space.norm       avgt   16  257390631.752 ± 25286960.868    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Old_Gen               avgt   16          0.088 ±        0.064  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Old_Gen.norm          avgt   16      28780.053 ±    21305.678    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Perm_Gen              avgt   16         ≈ 10⁻⁴                 MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Perm_Gen.norm         avgt   16         27.655 ±       20.636    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Survivor_Space        avgt   16          7.131 ±        0.404  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Survivor_Space.norm   avgt   16    2297223.339 ±   263432.586    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.count                          avgt   16        144.000                 counts
[info] ESIndexBenchmark.indexPhrase:·gc.time                           avgt   16       4072.000                     ms

[info] ESSuggestBenchmark.suggest         avgt   16  1538.747 ± 317.999  ms/op
```
previously
```
[info] Benchmark                                                                Mode  Cnt    Score    Error  Units
[info] c.s.s.common.CompletionAnalyzerBenchmark.analyzeCompletionTokens         avgt   16  114.496 ±  3.544  us/op
[info] c.s.s.secondary.ESIndexBenchmark.indexKeyword                            avgt   16  162.774 ±  6.349  ms/op
[info] c.s.s.secondary.ESIndexBenchmark.indexPhrase                             avgt   16  202.855 ± 25.170  ms/op

[info] ESIndexBenchmark.indexKeyword                                   avgt   16        140.958 ±       19.971   ms/op
[info] ESIndexBenchmark.indexKeyword:·gc.alloc.rate                    avgt   16        891.021 ±       99.848  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.alloc.rate.norm               avgt   16  317288910.442 ±  1659707.762    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Eden_Space           avgt   16        946.922 ±       24.514  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Eden_Space.norm      avgt   16  342537299.513 ± 54377844.178    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Old_Gen              avgt   16          0.227 ±        0.645  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Old_Gen.norm         avgt   16      78701.188 ±   219535.521    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Perm_Gen             avgt   16         ≈ 10⁻⁴                 MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Perm_Gen.norm        avgt   16         29.667 ±       17.032    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Survivor_Space       avgt   16          3.841 ±        0.391  MB/sec
[info] ESIndexBenchmark.indexKeyword:·gc.churn.PS_Survivor_Space.norm  avgt   16    1392249.296 ±   278919.764    B/op
[info] ESIndexBenchmark.indexKeyword:·gc.count                         avgt   16        144.000                 counts
[info] ESIndexBenchmark.indexKeyword:·gc.time                          avgt   16       4157.000                     ms
[info] ESIndexBenchmark.indexPhrase                                    avgt   16        164.295 ±       27.527   ms/op
[info] ESIndexBenchmark.indexPhrase:·gc.alloc.rate                     avgt   16        869.184 ±      106.271  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.alloc.rate.norm                avgt   16  355961273.045 ±  1372824.493    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Eden_Space            avgt   16        888.992 ±      181.758  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Eden_Space.norm       avgt   16  369787087.360 ± 87935589.692    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Old_Gen               avgt   16          0.058 ±        0.018  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Old_Gen.norm          avgt   16      24875.937 ±    10498.224    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Perm_Gen              avgt   16         ≈ 10⁻⁴                 MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Perm_Gen.norm         avgt   16         36.071 ±       30.033    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Survivor_Space        avgt   16          5.577 ±        0.413  MB/sec
[info] ESIndexBenchmark.indexPhrase:·gc.churn.PS_Survivor_Space.norm   avgt   16    2315335.304 ±   335743.104    B/op
[info] ESIndexBenchmark.indexPhrase:·gc.count                          avgt   16        145.000                 counts
[info] ESIndexBenchmark.indexPhrase:·gc.time                           avgt   16       4159.000                     ms

[info] ESSuggestBenchmark.suggest                 avgt   16  1191.591 ± 211.906  ms/op
```

[Review on Reviewable](https://reviewable.io/reviews/socrata-platform/spandex/69)

### CODE REVIEW

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexResult.scala [hunk 1]
```diff
@@ -33,8 +33,9 @@ object SpandexResult {
    * Transforms a search result with aggregation into spandex result options
    */
   def apply(response: SearchResults[FieldValue]): SpandexResult =
-    SpandexResult(response.aggs.map { src =>
-      SpandexOption(src.key, Some(src.docCount))
+    SpandexResult(response.thisPage.map { src =>
+      // TODO: aggregate with doc count
+      SpandexOption(src.value, Some(42))
```
Comment (sokrahta) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexResult.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1235292357/issues/source?bid=2110827&fileBranchId=2934517#l38)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala [hunk 2]
```diff
@@ -9,9 +9,9 @@ trait TestESData {
   val datasets = Seq("primus.1234", "primus.9876")
 
   def copies(dataset: String): Seq[DatasetCopy] = {
-    val snapshot    = DatasetCopy(dataset, 1, 5, LifecycleStage.Snapshotted) // scalastyle:ignore magic.number
-    val published   = DatasetCopy(dataset, 2, 10, LifecycleStage.Published) // scalastyle:ignore magic.number
-    val workingCopy = DatasetCopy(dataset, 3, 15, LifecycleStage.Unpublished) // scalastyle:ignore magic.number
+    val snapshot    = DatasetCopy(dataset, 1, 5, LifecycleStage.Snapshotted)
+    val published   = DatasetCopy(dataset, 2, 10, LifecycleStage.Published)
+    val workingCopy = DatasetCopy(dataset, 3, 15, LifecycleStage.Unpublished)
```
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1241284691/issues/source?bid=2110827&fileBranchId=2934517#l14)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala [hunk 3]
```diff
@@ -9,9 +9,9 @@ trait TestESData {
   val datasets = Seq("primus.1234", "primus.9876")
 
   def copies(dataset: String): Seq[DatasetCopy] = {
-    val snapshot    = DatasetCopy(dataset, 1, 5, LifecycleStage.Snapshotted) // scalastyle:ignore magic.number
-    val published   = DatasetCopy(dataset, 2, 10, LifecycleStage.Published) // scalastyle:ignore magic.number
-    val workingCopy = DatasetCopy(dataset, 3, 15, LifecycleStage.Unpublished) // scalastyle:ignore magic.number
+    val snapshot    = DatasetCopy(dataset, 1, 5, LifecycleStage.Snapshotted)
```
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1241284691/issues/source?bid=2110827&fileBranchId=2934517#l12)

File: spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala [hunk 4]
```diff
@@ -280,38 +281,56 @@ class SpandexElasticSearchClient(config: ElasticSearchConfig) extends ElasticSea
...
       .setSize(size)
-      .execute.actionGet
+
+    logElasticsearchRequest(search)
+    val response = search.execute.actionGet
+
     response.results[FieldValue](aggName)
   }
+
+  private def logElasticsearchRequest(request: SearchRequestBuilder): Unit = {
```
Comment (sokrahta) [inline spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala]: Codacy Issue found: [This value should have a smaller scope by using private[this]](https://www.codacy.com/app/johnkrah/spandex/file/1243418321/issues/source?bid=2110827&fileBranchId=2934517#l333)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 2 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1243418328/issues/source?bid=2110827&fileBranchId=2934517#l14)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 3 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1243418328/issues/source?bid=2110827&fileBranchId=2934517#l12)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 2 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1248683344/issues/source?bid=2110827&fileBranchId=2934517#l14)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 3 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1248683344/issues/source?bid=2110827&fileBranchId=2934517#l12)

**Comment by rlvoyer** [comment]
One thing we might consider for the future... While it's unlikely that we will change the underlying autocompletion algo. often, it would be nice to have some kind of accuracy metric attached to it, so that when we do make changes (such as this one), we can see how they're impacting the quality of the completions.

File: spandex-common/src/main/resources/reference.conf [hunk 5]
```diff
@@ -37,6 +41,17 @@ com.socrata.spandex {
...
+               "case_insensitive_word_prefix":{
+                 "tokenizer":"word",
+                 "filter":["lowercase","prefix"]
+               }
+             },
+             "filter":{
+               "prefix":{
+                 "type":"edgeNGram",
+                 "min_gram":1,
+                 "max_gram":32
```
Comment (rlvoyer) [inline spandex-common/src/main/resources/reference.conf]: How did we arrive at this max_ngram value?

File: spandex-perf/src/main/scala/com/socrata/spandex/common/ESSuggestBenchmark.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  val datasetId = "optimus.42"
+  var copyNumber = 1L
+  val systemColumnId = 2L
+  val userColumndId = "dead-beef"
+  val version = 3L
+  val stage = LifecycleStage.Published
+  val col = ColumnMap(datasetId, copyNumber, systemColumnId, userColumndId)
+  @Setup(Level.Iteration)
```
Comment (rlvoyer) [inline spandex-perf/src/main/scala/com/socrata/spandex/common/ESSuggestBenchmark.scala]: Insert newline

**Comment by rlvoyer** [comment]
LGTM

File: spandex-common/src/main/resources/reference.conf (hunk 5 above)
Comment (sokrahta) [inline spandex-common/src/main/resources/reference.conf]: carried over from the previous max analyzed shingle length.

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 2 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1273782156/issues/source?bid=2110827&fileBranchId=2934517#l14)

File: spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala (hunk 3 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/TestESData.scala]: Codacy Issue found: [Magic Number](https://www.codacy.com/app/johnkrah/spandex/file/1273782156/issues/source?bid=2110827&fileBranchId=2934517#l12)

**Comment by rlvoyer** [comment]
I reviewed in Reviewable, before once again noticing that Reviewable is for some unknown reason broken on this PR. Anyway, ship it. It looks good.

**Pull request merged**